solidity/ecdsa/contracts/libraries/Wallets.sol

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: MIT
+//
+// ▓▓▌ ▓▓ ▐▓▓ ▓▓▓▓▓▓▓▓▓▓▌▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▄
+// ▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▌▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
+//   ▓▓▓▓▓▓    ▓▓▓▓▓▓▓▀    ▐▓▓▓▓▓▓    ▐▓▓▓▓▓   ▓▓▓▓▓▓     ▓▓▓▓▓   ▐▓▓▓▓▓▌   ▐▓▓▓▓▓▓
+//   ▓▓▓▓▓▓▄▄▓▓▓▓▓▓▓▀      ▐▓▓▓▓▓▓▄▄▄▄         ▓▓▓▓▓▓▄▄▄▄         ▐▓▓▓▓▓▌   ▐▓▓▓▓▓▓
+//   ▓▓▓▓▓▓▓▓▓▓▓▓▓▀        ▐▓▓▓▓▓▓▓▓▓▓         ▓▓▓▓▓▓▓▓▓▓         ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
+//   ▓▓▓▓▓▓▀▀▓▓▓▓▓▓▄       ▐▓▓▓▓▓▓▀▀▀▀         ▓▓▓▓▓▓▀▀▀▀         ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▀
+//   ▓▓▓▓▓▓   ▀▓▓▓▓▓▓▄     ▐▓▓▓▓▓▓     ▓▓▓▓▓   ▓▓▓▓▓▓     ▓▓▓▓▓   ▐▓▓▓▓▓▌
+// ▓▓▓▓▓▓▓▓▓▓ █▓▓▓▓▓▓▓▓▓ ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  ▓▓▓▓▓▓▓▓▓▓
+// ▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓ ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  ▓▓▓▓▓▓▓▓▓▓
+//
+//                           Trust math, not hardware.
+
+pragma solidity ^0.8.9;
+
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+library Wallets {
+    using ECDSA for bytes32;
+
+    struct Wallet {
+        bytes32 publicKeyHash;
+        // Keccak256 hash of group members identifiers array. Group members do not
+        // include operators selected by the sortition pool that misbehaved during DKG.
+        // See how `misbehavedMembersIndices` are used in `hashGroupMembers` function.
+        bytes32 membersIdsHash;
+        bytes32 digestToSign;
+    }
+
+    struct Signature {
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+    }
+
+    struct Data {
+        // Mapping of keccak256 hashes of wallet public keys to wallet details.
+        mapping(uint256 => Wallet) registry;
+        // Holds keccak256 hashes of group public keys in the order of registration.
+        uint256 walletCounter;
+    }
+
+    event SignatureRequested(uint256 indexed walletID, bytes32 indexed digest);
+
+    event SignatureSubmitted(
+        uint256 indexed walletID,
+        bytes32 indexed digest,
+        Signature signature
+    );
+
+    /// @notice Registers a new wallet.
+    /// @param _membersIdsHash Keccak256 hash of group members identifiers array.
+    /// @param _publicKeyHash Keccak256 hash of group public key.
+    /// @return walletID ID of the newly registered wallet.
+    function addWallet(
+        Data storage self,
+        bytes32 _membersIdsHash,
+        bytes32 _publicKeyHash
+    ) internal returns (uint256 walletID) {
+        walletID = ++self.walletCounter;
```

Comment:
Why do we need `walletID`? Can't we use `_publicKeyHash` as key in the registry mapping? I think in other functions we can pass `publicKeyHash` instead of `walletID` and it should work as well.

Reply:
We would have to track `publicKeyHash` used as a key in the mapping similar as we do in the Random Beacon: https://github.com/keep-network/keep-core/blob/5138c7628868dbeed3ae2164f76fccc6c1fbb9e8/solidity/random-beacon/contracts/libraries/Groups.sol#L38

I thought that using the counter for wallets IDs will be cheaper.

Reply:
Right, but what I don't understand is why do we need to track it? When do we exactly need to know a `walletID` (as a counter) which can't be replaced by `publicKeyHash`? I might be wrong, but in random beacon we had a concept of group termination, expiration and for that purpose of calculating math we used an additional array.

Reply:
I simplified the implementation to use `publicKeyHash`. We will decide later how do we want to track wallets' lifecycle.

7011ba2f6816e0677e60b03bf13c2236817e4c69